orca-core/src/main/java/com/netflix/spinnaker/orca/DynamicTaskImplementationResolver.kt

```diff
@@ -0,0 +1,87 @@
+package com.netflix.spinnaker.orca
+
+import com.netflix.spinnaker.kork.dynamicconfig.DynamicConfigService
+import com.netflix.spinnaker.orca.api.pipeline.Task
+import com.netflix.spinnaker.orca.api.pipeline.graph.TaskNode.TaskDefinition
+import com.netflix.spinnaker.orca.api.pipeline.graph.TaskNode.DefinedTask
+import com.netflix.spinnaker.orca.api.pipeline.models.StageExecution
+import com.netflix.spinnaker.orca.api.pipeline.models.TaskExecution
+import com.netflix.spinnaker.orca.clouddriver.utils.CloudProviderAware
+import com.netflix.spinnaker.orca.config.TaskOverrideConfigurationProperties
+import com.netflix.spinnaker.orca.config.TaskOverrideConfigurationProperties.TaskOverrideDefinition
+import org.slf4j.LoggerFactory
+
+/**
+ * This resolver will modify the task definition if the original definition
+ * matches the criteria for task replacement and the configuration for such a replacement is enabled.
+ */
+class DynamicTaskImplementationResolver(
+  val dynamicConfigService: DynamicConfigService,
+  val taskOverrideConfigurationProperties: TaskOverrideConfigurationProperties?,
+  ): TaskImplementationResolver, CloudProviderAware {
+
+  private val log = LoggerFactory.getLogger(javaClass)
+
+  override fun resolve(stage: StageExecution, taskNode: DefinedTask): TaskExecution {
+    val taskOverrideDefinition = taskOverrideConfigurationProperties
+      ?.overrideDefinitions
+      ?.firstOrNull {
+        it.stageName.equals(stage.type, true) &&
+          it.originalTaskName.equals(taskNode.name, true)
```

Comment:
We should replace based on the the original task implementation class name.  The reason for this is the same task can have multiple names (look at the various names for `MonitorKatoTask` for example).